<style  type="text/css"> 
</style>

<b><center>
<span style="font-size: 24pt; line-height: 1.2">
COMS W4111-002 (Fall 2021)<br>Introduction to Databases
</span>
</center></b>
</span><br>
<p>
<i><center>
<span style="font-size: 20pt; line-height: 1.2">
Homework 1: Programming - 10 Points
</span>
</center></i>

__Note:__ Please replace the information below with your last name, first name and UNI.<br><br>



<i>
<span style="font-size: 20pt; line-height: 1.2"; >
AlShehri_Dana, da2975
</span>
</i>

## Introduction

### Objectives
This homework has you practice and build skill with:

- PART A: (1 point) Understanding relational databases
- PART B: (1 point) Understanding relational algebra
- PART C: (1 point) Cleaning data
- PART D: (1 point) Performing simple SQL queries to analyze the data.
- PART E: (6 points) CSVDataTable.py


__Note:__ The motivation for PART E may not be clear. The motivation will become clearer as the semester proceeds. The purpose of PART E is to get you started on programming in Python and manipulating data.



### Submission
1. File > Print Preview > Download as PDF
2. Upload .pdf and .ipynb to GradeScope
3. Upload CSVDataTable.py and CSVDataTable_Tests.py

**This assignment is due September 24, 11:59 pm ET**


### Collaboration
- You may use any information you get in TA or Prof. Ferguson's office hours, from lectures or from recitations.
- You may use information that you find on the web.
- You are NOT allowed to collaborate with other students outside of office hours.


# Part A: Written

1. What is a database management system?

    <i> A software tool that enables users to manage a meaningful collection of data in the form of a database for efficient use, accessibility, and security. </i>

2. What is a primary key and why is it important?

    <i> A primary key is an attribute belonging to an entity that has a unique value for each tuple such that no two or more tuples have the same value for that specific attribute </i>

3. Please explain the differences between SQL, MySQL Server and DataGrip?

    <i> SQL, which stands for Structured Query Language, is a language that's popularly used in relational databases to access and manipulate data within the database. <br>
    MySQL is a database management system that's supported in DataGrip, a database management environment that enables developers to work directly on databases, among other systems. <br>
 source = https://www.edureka.co/blog/what-is-mysql/ </i>

4. What are 4 different types of DBMS table relationships, give a brief explanaition for each?

    <i> Given two tables, table1 and table2, as reference examples, the following DBMS table relationships are possible:
1. One-to-One Relationship: table1 is associated with one attribute of table2 and vice versa.  <br>
2. One-to-Many Relationship: table1 is associated with one or more attributes of table2 and while table2 is only associated with one attribute of table1. <br>
3. Many-to-One Relationship: table1 is associated with only one attribute of table2 and while table2 is associated with  one or more attributes of table1 <br>
4. Many-to-Many Relationship: table1 is associated with one or more attributes of table2 and vice versa.  <br>
   <i>
    

5. What is an ER model? 

    <i> An ER model provides a visually detailed description of how entities are connected or associated with one another within a database design. </i>


6. Using Lucidchart draw an example of a logical ER model using Crow's Foot notation for Columbia classes. The entity types are:
    - Students, Professors, and Classes.
    - The relationships are:
        - A Class has exactly one Professor.
        - A Student has exactly one professor who is an _advisor._
        - A Professor may advise 0, 1 or many Students.
        - A Class has 0, 1 or many enrolled students.
        - A Student enrolls in 0, 1 or many Classes.
        
        
- You can define what you think are common attributes for each of the entity types. Do not define more than 5 or 6 attributes per entity type.


- In this example, explicitly show an example of a primary-key, foreign key, one-to-many relationship, and many-to-many relationship.


__Notes:__
- If you have not already done so, please register for a free account at Lucidchart.com. You can choose the option at the bottom of the left pane to add the ER diagram shapes.
- You can take a screen capture of you diagram and save in the zip directory that that contains you Jupyter notebook. Edit the following cell and replace "Boromir.jpg" with the name of the file containing your screenshot.
    
    

   <i> Use the following line to upload a photo of your Luicdchart. <i/>
   <img src="Logical_ER_Model.png">


Example of primary key: Student_ID is a primary key for Students.

Example of foreign key: Professor_ID in Students is a foreign key referencing Professor_ID of Professor. 

Example of one-to-many relationship: A professor advises 0, 1, or more students. A student has at most one professor as an advisor. 

Example of many-to-many relationship: 0, 1, or more students are enrolled in classes. 0, 1, or more classes are enrolled by students.

# Part B: Relational Algebra

You will use [the online relational calculator](https://dbis-uibk.github.io/relax/landing), choose the “Karlsruhe University of Applied Sciences” dataset.

An anti-join is a form of join with reverse logic. Instead of returning rows when there is a match (according to the join predicate) between the left and right side, an anti-join returns those rows from the left side of the predicate for which there is no match on the right. 

The Anti-Join Symbol is ▷.

Consider the following relational algebra expression and result.

/* (1) Set X = The set of classrooms in buildings Taylor or Watson. */

        X = σ building='Watson' ∨ building='Taylor' (classroom)

/* (2) Set Y =  The Anti-Join of department and X */

        Y = (department ▷ X)

/* (3) Display the rows in Y. */

        Y
        
   <img src="ra.png">


1. Find an alternate expression to (2) that computes the correct answer given X. Display the execution of your query below.

    <img src="query.png">
    <img src="ra2.png">
    <img src="ra2_result.png">


# Part C: Data Clean Up

## Please note: You MUST make a new schema using the lahmansdb_to_clean.sql file provided in the data folder. 

Use thelahmansdb_to_clean.sql file to make a new schema containing the raw data. The lahman database you created in Homework 0 has already been cleaned with all the constraints and will be used for Part D. Knowing how to clean data and add integrity constraints is very important which is why you go through the steps in part C. 

TLDR: If you use the HW0 lahman schema for this part you will get a lot of errors and recieve a lot of deductions. 

In [7]:
# You will need to follow instructions from HW 0 to make a new schema, import the data. 
# Connect to the unclean schema below by setting the database host, user ID and password.
%load_ext sql
%sql mysql+pymysql://root:56781234@localhost/lahmansdb_to_clean

The sql extension is already loaded. To reload it, use:
  %reload_ext sql



Data cleanup: For each table we want you to clean, we have provided a list of changes you have to make. You can reference the cleaned lahman db for inspiration and guidance, but know that there are different ways to clean the data and you will be graded for your choice rationalization. You should make these changes through DataGrip's workbench's table editor and/or using SQL queries. In this part you will clean two tables: People and Batting. 


### You must have: 
- A brief explanation of why you think the change we requested is important.
- What change you made to the table.
- Any queries you used to make the changes, either the ones you wrote or the Alter statements provided by DataGrip's editor. 
- Executed the test statements we provided
- The cleaned table's new create statement (after you finish all the changes) 


### Overview of Changes:

People Table 

   0. Primary Key (Explanation is given, but you still must add the key to your table yourself)
   1. Empty strings to NULLs
   2. Column typing
   3. isDead column
   4. deathDate and birthDate column

Batting Table

   1. Empty strings to NULLs
   2. Column typing
   3. Primary Key
   4. Foreign Key
   

   
### How to make the changes: 

__Using the Table Editor:__

When you hit apply, a popup will open displaying the ALTER statments sql generates. Copy the sql provided first and paste it into this notebook. Then you can apply the changes. This means that you are NOT executing the ALTER statements through your notebook. 


1. Right click on the table > Modify Table...
<img src="modify.png" width="400" height="800">

2. Keys > press the + button > input the parameters > Execute 
    OR Keys > press the + button > input the parameters > copy and paste the script generated under "SQL Script" and paste into your notebook > Run the cell in jupyter notebook
<img src="pk.png" width="600" height="1200">




__Using sql queries:__

Copy paste any queries that you write manually into the notebook as well!


<hr style="height:2px">

## People Table

### 0) EXAMPLE: Add a Primary Key 

(Solutions are given but make sure you still do this step in workbench!)

#### Explanation

We want to add a Primary Key because we want to be able to uniquely identify rows within our data. A primary key is also an index, which allows us to locate data faster. 

#### Change

I added a Primary Key on the playerID column and made the datatype VARCHAR(15)

__Note:__ This is for demonstration purposes only. playerID __is not__ a primary key for fielding.

#### SQL

~~~~sql
ALTER TABLE `lahmansdb_to_clean`.`people`
CHANGE COLUMN `playerID` `playerID` VARCHAR(15) NOT NULL ,
ADD PRIMARY KEY (`playerID`);
~~~~

#### Tests

In [10]:
%sql SHOW KEYS FROM people WHERE Key_name = 'PRIMARY'
    

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[('people', 0, 'PRIMARY', 1, 'playerID', 'A', 19894, None, None, '', 'BTREE', '', '', 'YES', None)]

### 1) Convert all empty strings to NULL

#### Explanation

Empty strings are considered to have a value in a database, while NULL indicates that there is no value in the database for that specific field. Hence, this step is important to avoid mistakes in retrieving information from the database. 

#### Change

I updated the table people of the lahmansdb_to_clean database by setting the empty strings found in the table to NULL

#### SQL

~~~~sql

update lahmansdb_to_clean.people
set birthYear=null
where birthYear='';


update lahmansdb_to_clean.people
set birthMonth=null
where birthMonth='';

update lahmansdb_to_clean.people
set birthDay=null
where birthDay='';

update lahmansdb_to_clean.people
set birthCountry=null
where birthCountry='';

update lahmansdb_to_clean.people
set birthState=null
where birthState='';

update lahmansdb_to_clean.people
set birthCity=null
where birthCity='';

update lahmansdb_to_clean.people
set deathYear=null
where deathYear='';

update lahmansdb_to_clean.people
set deathMonth=null
where deathMonth='';

update lahmansdb_to_clean.people
set deathDay=null
where deathDay='';

update lahmansdb_to_clean.people
set deathCountry=null
where deathCountry='';

update lahmansdb_to_clean.people
set deathState=null
where deathState='';

update lahmansdb_to_clean.people
set deathCity=null
where deathCity='';

update lahmansdb_to_clean.people
set nameFirst=null
where nameFirst='';

update lahmansdb_to_clean.people
set nameLast=null
where nameLast='';

update lahmansdb_to_clean.people
set nameGiven=null
where nameGiven='';

update lahmansdb_to_clean.people
set weight=null
where weight='';

update lahmansdb_to_clean.people
set height=null
where height='';

update lahmansdb_to_clean.people
set bats=null
where bats='';

update lahmansdb_to_clean.people
set throws=null
where throws='';

update lahmansdb_to_clean.people
set debut=null
where debut='';

update lahmansdb_to_clean.people
set finalGame=null
where finalGame='';

update lahmansdb_to_clean.people
set retroID=null
where retroID='';

update lahmansdb_to_clean.people
set bbrefID=null
where bbrefID='';
~~~~

#### Tests

In [11]:
%sql SELECT * FROM people WHERE birthState = ""


 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
0 rows affected.


[]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

Column datatypes must be changed to distinct appropriate values to ensure that its' value is interpreted in its' preferred data storage format.

#### Change

The datatypes I used for each column in the people table are as follows:  <br>
1. smallint unsigned datatype for birthYear, deathYear. <br>
2. tinyint datatype for birthMonth, birthDay, deathMonth, deathDay, height.  <br>
3. int datatype for weight. <br>
4. varchar datatype for birthCountry, birthState, birthCity, deathCountry, deathState, deathCity, nameFirst, nameLast, nameGiven, retroID, bbrefID.  <br>
5. char datatype for bats, throws. <br>
6. date datatype for debut, finalGame.

#### SQL

~~~~sql
alter table people modify birthYear smallint unsigned null;

alter table people modify birthMonth tinyint null;

alter table people modify birthDay tinyint null;

alter table people modify birthCountry varchar(30) null;

alter table people modify birthState varchar(30) null;

alter table people modify birthCity varchar(30) null;

alter table people modify deathYear smallint unsigned null;

alter table people modify deathMonth tinyint null;

alter table people modify deathDay tinyint null;

alter table people modify deathCountry varchar(30) null;

alter table people modify deathState varchar(30) null;

alter table people modify deathCity varchar(30) null;

alter table people modify nameFirst varchar(30) null;

alter table people modify nameLast varchar(30) null;

alter table people modify nameGiven varchar(50) null;

alter table people modify weight int null;

alter table people modify height tinyint null;

alter table people modify bats char(2) null;

alter table people modify throws char(2) null;

alter table people modify debut date null;

alter table people modify finalGame date null;

alter table people modify retroID varchar(15) null;

alter table people modify bbrefID varchar(15) null;

~~~~

### 3) Add an isDead Column that is either 'Y' or 'N'
- Some things to think of: What data type should this column be? How do you know if the player is dead or not? Maybe you do not know if the player is dead.
- You do not need to make guesses about life spans, etc. Just apply a simple rule.

'Y' means the player is dead

'N' means the player is alive

#### Explanation

Because of the data provided in the columns of the table, we can derive useful information and store it in one column. In this case, we can arrive to the conclusion of which players are still alive or dead.

#### Change

Given the known values in the DeathYear column, I compared the values with the current year (2021) to identify the players who have already died and set the value 'Y' in the corresponding isDead column data field. As for the unknown values, I set the value 'N' in their corresponding isDead column data field. 

#### SQL

~~~~sql
alter table lahmansdb_to_clean.people
add isDead char null;

update lahmansdb_to_clean.people
set isDead='Y'
where deathYear < 2021;

update lahmansdb_to_clean.people
set isDead='N'
where deathYear is null;
~~~~

#### Tests

In [6]:
%sql SELECT * FROM people WHERE isDead = "N" limit 10

 * mysql+pymysql://dbuser:***@localhost/lahman2019clean
10 rows affected.


[('aardsda01', '1981', '12', '27', 'USA', 'CO', 'Denver', None, None, None, None, None, None, 'David', 'Aardsma', 'David Allan', 215, '75', 'R', 'R', '2004-04-06', '2015-08-23', 'aardd001', 'aardsda01', datetime.datetime(1981, 12, 27, 0, 0), None, 'N'),
 ('aaronha01', '1934', '2', '5', 'USA', 'AL', 'Mobile', None, None, None, None, None, None, 'Hank', 'Aaron', 'Henry Louis', 180, '72', 'R', 'R', '1954-04-13', '1976-10-03', 'aaroh101', 'aaronha01', datetime.datetime(1934, 2, 5, 0, 0), None, 'N'),
 ('aasedo01', '1954', '9', '8', 'USA', 'CA', 'Orange', None, None, None, None, None, None, 'Don', 'Aase', 'Donald William', 190, '75', 'R', 'R', '1977-07-26', '1990-10-03', 'aased001', 'aasedo01', datetime.datetime(1954, 9, 8, 0, 0), None, 'N'),
 ('abadan01', '1972', '8', '25', 'USA', 'FL', 'Palm Beach', None, None, None, None, None, None, 'Andy', 'Abad', 'Fausto Andres', 184, '73', 'L', 'L', '2001-09-10', '2006-04-13', 'abada001', 'abadan01', datetime.datetime(1972, 8, 25, 0, 0), None, 'N'),
 ('abadfe01', '1985', '12', '17', 'D.R.', 'La Romana', 'La Romana', None, None, None, None, None, None, 'Fernando', 'Abad', 'Fernando Antonio', 220, '73', 'L', 'L', '2010-07-28', '2017-10-01', 'abadf001', 'abadfe01', datetime.datetime(1985, 12, 17, 0, 0), None, 'N'),
 ('abbotgl01', '1951', '2', '16', 'USA', 'AR', 'Little Rock', None, None, None, None, None, None, 'Glenn', 'Abbott', 'William Glenn', 200, '78', 'R', 'R', '1973-07-29', '1984-08-08', 'abbog001', 'abbotgl01', datetime.datetime(1951, 2, 16, 0, 0), None, 'N'),
 ('abbotje01', '1972', '8', '17', 'USA', 'GA', 'Atlanta', None, None, None, None, None, None, 'Jeff', 'Abbott', 'Jeffrey William', 190, '74', 'R', 'L', '1997-06-10', '2001-09-29', 'abboj002', 'abbotje01', datetime.datetime(1972, 8, 17, 0, 0), None, 'N'),
 ('abbotji01', '1967', '9', '19', 'USA', 'MI', 'Flint', None, None, None, None, None, None, 'Jim', 'Abbott', 'James Anthony', 200, '75', 'L', 'L', '1989-04-08', '1999-07-21', 'abboj001', 'abbotji01', datetime.datetime(1967, 9, 19, 0, 0), None, 'N'),
 ('abbotku01', '1969', '6', '2', 'USA', 'OH', 'Zanesville', None, None, None, None, None, None, 'Kurt', 'Abbott', 'Kurt Thomas', 180, '71', 'R', 'R', '1993-09-07', '2001-04-13', 'abbok002', 'abbotku01', datetime.datetime(1969, 6, 2, 0, 0), None, 'N'),
 ('abbotky01', '1968', '2', '18', 'USA', 'MA', 'Newburyport', None, None, None, None, None, None, 'Kyle', 'Abbott', 'Lawrence Kyle', 200, '76', 'L', 'L', '1991-09-10', '1996-08-24', 'abbok001', 'abbotky01', datetime.datetime(1968, 2, 18, 0, 0), None, 'N')]

### 4) Add a deathDate and birthDate column
Some things to think of: What do you do if you are missing information? What datatype should this column be?

You have to create this column from other columns in the table.

#### Explanation

It is more useful to include a column that represents the full date (i.e. year, month, day) rather than have it seperated in three columns. It simplifes information extraction when it's possible to retrieve data from one column rather than multiple columns.

#### Change

I added deathDate column to the table with data type 'date'. Then, I changed the deathDate value from null to year-month-day format. After that, I changed the data type of the column to 'datetime'. I did the same thing for birthDate.

#### SQL

~~~~sql
alter table lahmansdb_to_clean.people
	add deathDate date null;
    
update lahmansdb_to_clean.people
set deathDate=concat(deathYear,'-',deathMonth,'-',deathDay,'')
where deathDay or deathMonth or deathYear;

alter table lahmansdb_to_clean.people modify deathDate datetime null;

alter table lahmansdb_to_clean.people
	add birthDate date null;

update lahmansdb_to_clean.people
set people.birthDate=concat(birthYear,'-',birthMonth,'-',birthDay,'')
where birthDay or birthMonth or birthYear;

alter table lahmansdb_to_clean.people modify birthDate datetime null;
~~~~

#### Tests

In [12]:
%sql SELECT deathDate FROM people WHERE deathDate >= '2005-01-01' ORDER BY deathDate ASC LIMIT 10;

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
10 rows affected.


[(datetime.datetime(2005, 1, 4, 0, 0),),
 (datetime.datetime(2005, 1, 7, 0, 0),),
 (datetime.datetime(2005, 1, 9, 0, 0),),
 (datetime.datetime(2005, 1, 10, 0, 0),),
 (datetime.datetime(2005, 1, 21, 0, 0),),
 (datetime.datetime(2005, 1, 22, 0, 0),),
 (datetime.datetime(2005, 1, 31, 0, 0),),
 (datetime.datetime(2005, 2, 4, 0, 0),),
 (datetime.datetime(2005, 2, 8, 0, 0),),
 (datetime.datetime(2005, 2, 11, 0, 0),)]

In [13]:
%sql SELECT birthDate FROM people WHERE birthDate <= '1965-01-01' ORDER BY birthDate ASC LIMIT 10;


 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
10 rows affected.


[(datetime.datetime(1820, 4, 17, 0, 0),),
 (datetime.datetime(1824, 10, 5, 0, 0),),
 (datetime.datetime(1832, 9, 17, 0, 0),),
 (datetime.datetime(1832, 10, 23, 0, 0),),
 (datetime.datetime(1835, 1, 10, 0, 0),),
 (datetime.datetime(1836, 2, 29, 0, 0),),
 (datetime.datetime(1837, 12, 26, 0, 0),),
 (datetime.datetime(1838, 3, 10, 0, 0),),
 (datetime.datetime(1838, 7, 16, 0, 0),),
 (datetime.datetime(1838, 8, 27, 0, 0),)]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
create table lahmansdb_to_clean.people
(
	playerID varchar(15) not null
		primary key,
	birthYear smallint unsigned null,
	birthMonth tinyint null,
	birthDay tinyint null,
	birthCountry varchar(30) null,
	birthState varchar(30) null,
	birthCity varchar(30) null,
	deathYear smallint unsigned null,
	deathMonth tinyint null,
	deathDay tinyint null,
	deathCountry varchar(30) null,
	deathState varchar(30) null,
	deathCity varchar(30) null,
	nameFirst varchar(30) null,
	nameLast varchar(30) null,
	nameGiven varchar(50) null,
	weight int null,
	height tinyint null,
	bats char(2) null,
	throws char(2) null,
	debut date null,
	finalGame date null,
	retroID varchar(15) null,
	bbrefID varchar(15) null,
	isDead char null,
	deathDate datetime null,
	birthDate datetime null
);

~~~~

<hr style="height:2px">

## Batting Table

### 1) Convert all empty strings to NULL

#### Explanation

Empty strings are considered to have a value in a database, while NULL indicates that there is no value in the database for that specific field. Hence, this step is important to avoid mistakes in retrieving information from the database. 

#### Change

I updated the table batting of the lahmansdb_to_clean database by setting the empty strings found in the table to NULL

#### SQL

~~~~sql
update lahmansdb_to_clean.batting
set yearID=null
where yearID='';

update lahmansdb_to_clean.batting
set stint=null
where stint='';

update lahmansdb_to_clean.batting
set teamID=null
where teamID='';

update lahmansdb_to_clean.batting
set lgID=null
where lgID='';

update lahmansdb_to_clean.batting
set G=null
where G='';

update lahmansdb_to_clean.batting
set AB=null
where AB='';

update lahmansdb_to_clean.batting
set R=null
where R='';

update lahmansdb_to_clean.batting
set H=null
where H='';

update lahmansdb_to_clean.batting
set 2B=null
where 2B='';

update lahmansdb_to_clean.batting
set 3B=null
where 3B='';

update lahmansdb_to_clean.batting
set HR=null
where HR='';

update lahmansdb_to_clean.batting
set RBI=null
where RBI='';

update lahmansdb_to_clean.batting
set SB=null
where SB='';

update lahmansdb_to_clean.batting
set CS=null
where CS='';

update lahmansdb_to_clean.batting
set BB=null
where BB='';

update lahmansdb_to_clean.batting
set SO=null
where SO='';

update lahmansdb_to_clean.batting
set IBB=null
where IBB='';

update lahmansdb_to_clean.batting
set HBP=null
where HBP='';

update lahmansdb_to_clean.batting
set SH=null
where SH='';

update lahmansdb_to_clean.batting
set SF=null
where SF='';

update lahmansdb_to_clean.batting
set GIDP=null
where GIDP='';
~~~~

#### Tests

In [15]:
%sql SELECT count(*) FROM lahmansdb_to_clean.batting where RBI is NULL;

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[(756,)]

### 2) Change column datatypes to appropriate values (ENUM, INT, VARCHAR, DATETIME, ETC)

#### Explanation

Column datatypes must be changed to distinct appropriate values to ensure that its' value is interpreted in its' preferred data storage format.

#### Change

The datatypes I used for each column in the batting table are as follows:  <br>
1. smallint datatype for yearID, G,AB, R, H, 2B, 3B, HR, RBI, SB, CS, BB, SO, IBB, HBP, SH, SF, GIDP <br>
2. tinyint datatype for stint,  <br>
3. varchar datatype for playerID.  <br>
4. char datatype for teamID, lgID.

#### SQL

~~~~sql
alter table batting modify playerID varchar(10) null;

alter table batting modify yearID smallint null;

alter table batting modify stint tinyint null;

alter table batting modify teamID char(3) null;

alter table batting modify lgID char(2) null;

alter table batting modify G smallint null;

alter table batting modify AB smallint null;

alter table batting modify R smallint null;

alter table batting modify H smallint null;

alter table batting modify `2B` smallint null;

alter table batting modify `3B` smallint null;

alter table batting modify HR smallint null;

alter table batting modify RBI smallint null;

alter table batting modify SB smallint null;

alter table batting modify CS smallint null;

alter table batting modify BB smallint null;

alter table batting modify SO smallint null;

alter table batting modify IBB smallint null;

alter table batting modify HBP smallint null;

alter table batting modify SH smallint null;

alter table batting modify SF smallint null;

alter table batting modify GIDP smallint null;
~~~~

### 3) Add a Primary Key
Two options for the Primary Key:

- Composite Key: playerID, yearID, stint
- Covering Key (Index): playerID, yearID, stint, teamID


#### Choice

Composite Key: playerID, yearID, stint

#### Explanation


Sometimes two or more columns make a unique combination in comparision to them individually. As such, using a composite key in the batting table ensures we can uniquely identify each row in the table. 

#### SQL

~~~~sql
alter table batting
	add constraint batting_pk
		primary key (playerID, yearID, stint);
~~~~

#### Test

In [16]:
%sql SHOW KEYS FROM batting WHERE Key_name = 'PRIMARY' and Column_name = 'playerID'

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[('batting', 0, 'PRIMARY', 1, 'playerID', 'A', 19582, None, None, '', 'BTREE', '', '', 'YES', None)]

### 4) Add a foreign key on playerID between the People and Batting Tables
Note: Two people in the batting table do not exist in the people table. How should you handle this issue?

#### Explanation

People in the batting table can be related to people in the people table by means of having a foreign key column from one table (i.e. batting table) with values that correspond to the values of the primary key column of the other table (i.e. people table). This enables the data linkage between the two tables.  

#### Change

I added a foreign key to the batting table that references the primary key of the people table

#### SQL

~~~~sql
alter table batting
	add constraint batting_people_playerID_fk
		foreign key (playerID) references people (playerID);
~~~~

#### Tests

In [17]:
%sql Select playerID from batting where playerID not in (select playerID from people);

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
0 rows affected.


[]

### Final CREATE Statement

To find the create statement:

- Right click on the table name in workbench
- Select 'Copy to Clipboard'
- Select 'Create Statement' 

The create statement will now be copied into your clipboard and can be pasted into the cell below. 

~~~~sql
create table lahmansdb_to_clean.batting
(
	playerID varchar(10) not null,
	yearID smallint not null,
	stint tinyint not null,
	teamID char(3) null,
	lgID char(2) null,
	G smallint null,
	AB smallint null,
	R smallint null,
	H smallint null,
	`2B` smallint null,
	`3B` smallint null,
	HR smallint null,
	RBI smallint null,
	SB smallint null,
	CS smallint null,
	BB smallint null,
	SO smallint null,
	IBB smallint null,
	HBP smallint null,
	SH smallint null,
	SF smallint null,
	GIDP smallint null,
	primary key (playerID, yearID, stint),
	constraint batting_people_playerID_fk
		foreign key (playerID) references lahmansdb_to_clean.people (playerID)
);

~~~~

# Part D: SQL Queries

NOTE: You must use the CLEAN lahman schema provided in HW0 for the queries below to ensure your answers are consistent with the solutions. 

In [19]:
%reload_ext sql
%sql mysql+pymysql://root:56781234@localhost/lahmansbaseballdb

## Question 0
What is the average salary in baseball history?

In [24]:
%sql select avg(salary) from lahmansbaseballdb.salaries;

 * mysql+pymysql://root:***@localhost/lahmansbaseballdb
   mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[(2085634.053125473,)]

## Question 1
Select the players with a first name of Sam who were born in the United States and attended college.

Include their first name, last name, playerID, school ID, yearID and birth state. Limit 10

Hint: Use a Join between People and CollegePlaying

In [33]:
%%sql select nameFirst, nameLast, people.playerID, schoolID, yearID, birthState
from (lahmansbaseballdb.people join lahmansbaseballdb.collegeplaying on people.playerID = collegeplaying.playerID)
where nameFirst='Sam' and birthCountry='USA'and people.playerID=collegeplaying.playerID limit 10;

 * mysql+pymysql://root:***@localhost/lahmansbaseballdb
   mysql+pymysql://root:***@localhost/lahmansdb_to_clean
10 rows affected.


[('Sam', 'Barnes', 'barnesa01', 'auburn', 1918, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1919, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1920, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1921, 'AL'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1956, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1957, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1958, 'NC'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1971, 'GA'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1972, 'GA'),
 ('Sam', 'Brown', 'brownsa01', 'grovecity', 1899, 'PA')]

In [3]:
%%sql 

 * mysql+pymysql://root:***@localhost/lahmansbaseballdb
10 rows affected.


[('Sam', 'Barnes', 'barnesa01', 'auburn', 1918, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1919, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1920, 'AL'),
 ('Sam', 'Barnes', 'barnesa01', 'auburn', 1921, 'AL'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1956, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1957, 'NC'),
 ('Sam', 'Bowens', 'bowensa01', 'tennst', 1958, 'NC'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1971, 'GA'),
 ('Sam', 'Bowen', 'bowensa02', 'gacoast', 1972, 'GA'),
 ('Sam', 'Brown', 'brownsa01', 'grovecity', 1899, 'PA')]

## Question 2
Update all entries with full_name Columbia University to 'Columbia University in the City of New York' in the Schools table. Then select the row.


In [ ]:
%sql update lahmansbaseballdb.schools 
set name_full='Columbia University in the City of New York' 
where name_full='Columbia University';

In [44]:
%sql select * from lahmansbaseballdb.schools where schoolID='columbia';

 * mysql+pymysql://root:***@localhost/lahmansbaseballdb
   mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[('columbia', 'Columbia University in the City of New York', 'New York', 'NY', 'USA')]

 * mysql+pymysql://root:***@localhost/lahmansdb_to_clean
1 rows affected.


[('columbia', 'Columbia University in the City of New York', 'New York', 'NY', 'USA')]

# Part E: CSVDataTable


## i. Conceptual Questions
The purpose of this homework is to teach you the behaviour of SQL Databases by asking you to implement functions that will model the behaviour of a real database with CSVDataTable. You will mimic a SQL Database using CSV files. 

Read through the scaffolding code provided in the CSVDataTable folder first to understand and answer the following conceptual questions.

1. Given this SQL statement:
    
        SELECT nameFirst, nameLast FROM people WHERE playerID = collied01
   
   If you run find_by_primary_key() on this statement, what are key_fields and field_list? 

    <i> key_fields is the value "collied01" while field_list are the columns nameFirst and nameLast. </i>

2. What should be checked when you are trying to INSERT a new row into a table with a PK?

    <i> There should not be two rows with identical PK. Hence, it should be checked that the new row to be inserted does not have a PK value that already exists in the table or that is null. </i>

3. What should be checked when you are trying to UPDATE a row in a table with a PK?

    <i> It should be checked if the row exists in the table and if it is up to date or not. </i>

## ii. Coding


You are responsible for implementing and testing two classes in Python: CSVDataTable, BaseDataTable.
The python files and data can be found in the assignment under Courseworks. 

We have already given you **find_by_template(self, template, field_list=None, limit=None, offset=None, order_by=None)**
Use this as a jumping off point for the rest of your functions.

Methods to complete:

CSVDataTable.py
- find_by_primary_key(self, key_fields, field_list=None)
- delete_by_key(self, key_fields)
- delete_by_template(self, template)
- update_by_key(self, key_fields, new_values)
- update_by_template(self, template, new_values)
- insert(self, new_record)
CSV_table_tests.py
- You must test all methods. You will have to write these tests yourself. 
- You must test your methods on the People and Batting table.

If you do not include tests and tests outputs 50% of this section's points will be deducted at the start




## iii. Testing

Please copy the text from the output of your tests and paste it below:


find_by_primary_key(): Known Record
{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}

find_by_primary_key(): Unknown Record
Unknown record: No record found

find_by_template(): Known Template
[{'playerID': 'aardsda01', 'birthYear': '1981', 'birthMonth': '12', 'birthDay': '27', 'birthCountry': 'USA', 'birthState': 'CO', 'birthCity': 'Denver', 'deathYear': '', 'deathMonth': '', 'deathDay': '', 'deathCountry': '', 'deathState': '', 'deathCity': '', 'nameFirst': 'David', 'nameLast': 'Aardsma', 'nameGiven': 'David Allan', 'weight': '215', 'height': '75', 'bats': 'R', 'throws': 'R', 'debut': '2004-04-06', 'finalGame': '2015-08-23', 'retroID': 'aardd001', 'bbrefID': 'aardsda01'}]

find_by_template(): Unknown Template
[]

delete_by_key(): Known Record
Number of rows deleted = 1

delete_by_key(): Unknown Record
Unknown record: No record found

delete_by_template(): Known Template
Number of rows deleted = 3

delete_by_template(): Unknown Template
Number of rows deleted = 0

update_by_key(): Known Record
Number of rows updated = 1

update_by_key(): Unknown Record
Unknown record: No record found

update_by_template(): Known Template
Number of rows updated = 3

update_by_template(): Unknown Template
Unknown template: No results found

insert(): New Record with PK
None

insert(): New Record with PK that already exists
Error: primary key already exists


find_by_primary_key(): Known Record
{'playerID': 'abercda01', 'yearID': '1871', 'stint': '1', 'teamID': 'TRO', 'lgID': 'NA', 'G': '1', 'AB': '4', 'R': '0', 'H': '0', '2B': '0', '3B': '0', 'HR': '0', 'RBI': '0', 'SB': '0', 'CS': '0', 'BB': '0', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}

find_by_primary_key(): Unknown Record
Unknown record: No record found

find_by_template(): Known Template
[{'playerID': 'addybo01', 'yearID': '1871', 'stint': '1', 'teamID': 'RC1', 'lgID': 'NA', 'G': '25', 'AB': '118', 'R': '30', 'H': '32', '2B': '6', '3B': '0', 'HR': '0', 'RBI': '13', 'SB': '8', 'CS': '1', 'BB': '4', 'SO': '0', 'IBB': '', 'HBP': '', 'SH': '', 'SF': '', 'GIDP': '0'}]

find_by_template(): Unknown Template
[]

delete_by_key(): Known Record
Number of rows deleted = 1

delete_by_key(): Unknown Record
Unknown record: No record found

delete_by_template(): Known Template
Number of rows deleted = 11

delete_by_template(): Unknown Template
Number of rows deleted = 0

update_by_key(): Known Record
Number of rows updated = 1

update_by_key(): Unknown Record
Unknown record: No record found

update_by_template(): Known Template
Number of rows updated = 4

update_by_template(): Unknown Template
Unknown template: No results found

insert(): New Record with PK
None

insert(): New Record with PK that already exists
Error: primary key already exists

Process finished with exit code 0